plugins/api/src/switch_user.cpp

```diff
@@ -0,0 +1,290 @@
+#include "irods/plugins/api/api_plugin_number.h"
+#include "irods/plugins/api/switch_user_types.h"
+#include "irods/rodsDef.h"
+#include "irods/rcConnect.h"
+#include "irods/rodsPackInstruct.h"
+#include "irods/rcMisc.h"
+#include "irods/client_api_allowlist.hpp"
+
+#include "irods/apiHandler.hpp"
+
+#include <functional>
+
+#ifdef RODS_SERVER
+
+//
+// Server-side Implementation
+//
+
+// clang-format off
+#include "irods/switch_user.h"
+#include "irods/irods_rs_comm_query.hpp"
+#include "irods/rodsErrorTable.h"
+#include "irods/irods_logger.hpp"
+#include "irods/irods_server_properties.hpp"
+#include "irods/server_utilities.hpp"
+#include "irods/version.hpp"
+
+#define IRODS_USER_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+#include "irods/user_administration.hpp"
+
+#include <cstring>
+#include <string>
+#include <string_view>
+// clang-format on
+
+// This global variable is required so that the API plugin can access RcComm connections
+// created as a result of previous API calls (i.e. connections created due to redirection).
+extern zoneInfo* ZoneInfoHead;
+
+namespace
+{
+    //
+    // Namespace and Type aliases
+    //
+
+    namespace adm = irods::experimental::administration;
+
+    using log_api = irods::experimental::log::api;
+
+    //
+    // Function Prototypes
+    //
+
+    auto call_switch_user(irods::api_entry*, RsComm*, SwitchUserInput*) -> int;
+
+    auto rs_switch_user(RsComm*, SwitchUserInput*) -> int;
+
+    auto check_input(const SwitchUserInput*) -> int;
+
+    auto update_user_info(UserInfo&, const char*, const char*, const char*) -> void;
+
+    auto update_privilege_level(AuthInfo&, adm::user_type, const std::string_view, const std::string_view) -> void;
+
+    auto update_server_to_server_connections(RsComm&) -> void;
+
+    //
+    // Function Implementations
+    //
+
+    auto call_switch_user(irods::api_entry* _api, RsComm* _comm, SwitchUserInput* _input) -> int
+    {
+        return _api->call_handler<SwitchUserInput*>(_comm, _input);
+    } // call_switch_user
+
+    auto check_input(const SwitchUserInput* _input) -> int
+    {
+        if (!_input) {
+            log_api::error("Invalid input argument: received a null pointer");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        if (std::strlen(_input->username) == 0) {
+            log_api::error("Invalid username argument: empty string");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        if (std::strlen(_input->zone) == 0) {
+            log_api::error("Invalid zone argument: empty string");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        return 0;
+    } // check_input
```

Comment:
The input validation performed in the rc function need to be performed here.

Consider removing some of the checks from the rc function if it makes sense.

Reply:
done